PickAtRandom bean processor for locking one item in a batch

paterczm opened the pull request.

Commit:
PickAtRandom bean processor for locking one item in a batch

bserdar commented:
This can be improved a bit, maybe with a callback, or by changing this bean to an abstract base. Given an array of records, it can pick at random, and call the callback/abstract method to see if the picked element is acceptable. If not, it picks another one until the array is exhausted or one of the elements is accepted. The random selection must select each element only once. The callback/abstract method implementation would perform the actual locking.

Commit:
If lock is already taken, try to lock another element in the batch without querying for the batch again

paterczm commented:
The way it was implemented it would query for a new batch of elements each time lock acquire was unsuccessful. I agree this is not optimal, rewrote this logic to try to lock until it runs out of elements in the batch - only then it will stop processing and query for another batch.

Also removed PingPredicate. We are not using it right now and I would need to spend some time to make it work after the refactorings I did (refactorings to use dependency injection rather than camel expressions).

bserdar commented on src/main/java/com/redhat/lightblue/camel/lock/LightblueLockPolicy.java:
```diff
@@ -56,57 +65,114 @@ public void beforeWrap(RouteContext routeContext, ProcessorDefinition<?> definit
-                    LOGGER.debug("About to aquire lock on "+ java.net.URLDecoder.decode(resourceId, "UTF-8"));
+                if (callerId == null)
+                    locking.setCallerId(URLEncoder.encode(routeContext.getRoute().getId()+"-"+Thread.currentThread().getId(), "UTF-8"));
 
```
Are you sure this is a unique Id when there are multiple hosts involved?

paterczm replied:
Nope. Good catch.

bserdar commented on src/main/java/com/redhat/lightblue/camel/lock/LightblueLockPolicy.java:
```diff
@@ -56,57 +65,114 @@ public void beforeWrap(RouteContext routeContext, ProcessorDefinition<?> definit
+                return new ImmutablePair<T, String>(element, resourceId);
+            } else if (elements.length > 1) {
+                T[] newElements = ArrayUtils.remove(elements, index);
+                return tryToLock(newElements);
```
Why recursion, and not loop?

paterczm replied:
Recursion is just more understandable / readable to me. Any problem with recursion?

bserdar commented on src/main/java/com/redhat/lightblue/camel/lock/LightblueLockPolicy.java:
```diff
@@ -56,57 +65,114 @@ public void beforeWrap(RouteContext routeContext, ProcessorDefinition<?> definit
+            if (locking.acquire(resourceId, ttl)) {
+                return new ImmutablePair<T, String>(element, resourceId);
+            } else if (elements.length > 1) {
+                T[] newElements = ArrayUtils.remove(elements, index);
```
According to javadocs, this copies the input array. An unsuccessful locking attempt for an array of 100 will result 100 copies of the array with successively fewer elements.

A better approach here would be to add all the elements to an array list on method entry, and remove them during processing. Less memory intensive, easy to convert to loop.

paterczm replied:
How much of a problem do you think that is? Completely unsuccessful locking (it was not possible to lock a single element) will be very rare. And GC should not have problems dealing with that.

bserdar replied:
The GC can deal with that only after all of the recursions return.

It is not a problem of memory. It is simply inefficient. I'll leave it to
you.

On Mon, Oct 26, 2015 at 12:31 PM, Marek notifications@github.com wrote:

> In src/main/java/com/redhat/lightblue/camel/lock/LightblueLockPolicy.java
> https://github.com/lightblue-platform/lightblue-camel/pull/43#discussion_r43033367
> :
> 
> > -            String resourceId = resourceIdExtractor.getResourceId(element);
> >   +
> > -            if (LOGGER.isDebugEnabled())
> > -                LOGGER.debug("About to aquire lock on "+ java.net.URLDecoder.decode(resourceId, "UTF-8"));
> >   +
> > -            if (locking.acquire(resourceId, ttl)) {
> > -                return new ImmutablePair<T, String>(element, resourceId);
> > -            } else if (elements.length > 1) {
> > -                T[] newElements = ArrayUtils.remove(elements, index);
> 
> How much of a problem do you think that is? Completely unsuccessful
> locking (it was not possible to lock a single element) will be very rare.
> And GC should not have problems dealing with that.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/lightblue-platform/lightblue-camel/pull/43/files#r43033367
> .

Commit:
Use random callerId to avoid locking conflicts between jvms

Commit:
Using list instead of an array for better performance when removing elements

Commit:
Using a loop instead of recursion in tryToLock
